### ISSUE TITLE
Flutter syncfusion DataGrid on canSubmitCell event call new page to choose value. where updown and tab keys are not working

### ISSUE THREAD

**Issue by abdulrehmananwar** [syncfusion/flutter-widgets#1073]
### **Issue when i navigate to new page arrow up down key and tab key not work to new page. i calling new page from canSubmitCell event.** 
Below is Code.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:syncfusion_flutter_datagrid/datagrid.dart';
import 'package:collection/collection.dart';

void main() {
  runApp(const MaterialApp(home: SfDataGridDemo()));
}

class SfDataGridDemo extends StatefulWidget {
  const SfDataGridDemo({Key? key}) : super(key: key);

  @override
  SfDataGridDemoState createState() => SfDataGridDemoState();
}

DataGridController dataGridController = DataGridController();

class SfDataGridDemoState extends State<SfDataGridDemo> {
  List<Employee> _employees = <Employee>[];
  late EmployeeDataSource _employeeDataSource;

  @override
  void initState() {
    super.initState();
    _employees = getEmployeeData();
    _employeeDataSource = EmployeeDataSource(_employees,this.context);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Syncfusion Flutter DataGrid')),
      body: SfDataGrid(
          source: _employeeDataSource,
          controller: dataGridController,
          columns: getColumns,
          allowEditing: true,
          navigationMode: GridNavigationMode.cell,
          selectionMode: SelectionMode.single,
          editingGestureType: EditingGestureType.tap,
          columnWidthMode: ColumnWidthMode.fill),
    );
  }

  List<GridColumn> get getColumns {
    return <GridColumn>[
      GridColumn(
          columnName: 'id',
          label: Container(
              padding: const EdgeInsets.symmetric(horizontal: 16.0),
              alignment: Alignment.centerRight,
              child: const Text('ID'))),
      GridColumn(
          columnName: 'name',
          label: Container(
              padding: const EdgeInsets.symmetric(horizontal: 16.0),
              alignment: Alignment.centerLeft,
              child: const Text('Name'))),
      GridColumn(
          columnName: 'designation',
          label: Container(
              padding: const EdgeInsets.symmetric(horizontal: 16.0),
              alignment: Alignment.centerLeft,
              child:
              const Text('Designation', overflow: TextOverflow.ellipsis))),
      GridColumn(
          columnName: 'salary',
          label: Container(
              padding: const EdgeInsets.symmetric(horizontal: 16.0),
              alignment: Alignment.centerRight,
              child: const Text('Salary'))),
    ];
  }

  List<Employee> getEmployeeData() {
    return [
      Employee(10001, 'Lara', 'Project Lead', 45000),
      Employee(10002, 'James', 'Developer', 23000),
      Employee(10003, 'Kathryn', 'Project Lead', 43500),
      Employee(10004, 'Michael', 'Manager', 50000),
      Employee(10005, 'Martin', 'Developer', 25000),
      Employee(10006, 'Jack', 'Project Lead', 54000),
      Employee(10007, 'Balnc', 'UI Designer', 28500),
      Employee(10008, 'Lara', 'Project Lead', 40000),
      Employee(10009, 'James', 'Developer', 23000),
      Employee(10010, 'Kathryn', 'Project Lead', 43500)
    ];
  }
}

class EmployeeDataSource extends DataGridSource {
  EmployeeDataSource(this.employees,this.context) {
    dataGridRows = employees
        .map<DataGridRow>((dataGridRow) => dataGridRow.getDataGridRow())
        .toList();
  }

  BuildContext context;
  List<Employee> employees = [];
  List<DataGridRow> dataGridRows = [];

  bool isSearching = false;
  bool valueChange = false;

  /// Helps to hold the new value of all editable widget.
  /// Based on the new value we will commit the new value into the corresponding
  /// [DataGridCell] on [onSubmitCell] method.
  dynamic newCellValue;

  /// Help to control the editable text in [TextField] widget.
  TextEditingController editingController = TextEditingController();

  @override
  List<DataGridRow> get rows => dataGridRows;

  @override
  DataGridRowAdapter? buildRow(DataGridRow row) {
    return DataGridRowAdapter(
        cells: row.getCells().map<Widget>((dataGridCell) {
          return Container(
              alignment: (dataGridCell.columnName == 'id' ||
                  dataGridCell.columnName == 'salary')
                  ? Alignment.centerRight
                  : Alignment.centerLeft,
              padding: const EdgeInsets.symmetric(horizontal: 16.0),
              child: Text(
                dataGridCell.value.toString(),
                overflow: TextOverflow.ellipsis,
              ));
        }).toList());
  }

  @override
  void onCellSubmit(DataGridRow dataGridRow, RowColumnIndex rowColumnIndex,
      GridColumn column) {
    final dynamic oldValue = dataGridRow
        .getCells()
        .firstWhereOrNull((DataGridCell dataGridCell) =>
    dataGridCell.columnName == column.columnName)
        ?.value ??
        '';

    final int dataRowIndex = dataGridRows.indexOf(dataGridRow);

    if (newCellValue == null || oldValue == newCellValue) {
      return;
    }

    if (column.columnName == 'id') {
      dataGridRows[dataRowIndex].getCells()[rowColumnIndex.columnIndex] =
          DataGridCell<int>(columnName: 'id', value: newCellValue);
      employees[dataRowIndex].id = newCellValue as int;
    } else if (column.columnName == 'name') {
      dataGridRows[dataRowIndex].getCells()[rowColumnIndex.columnIndex] =
          DataGridCell<String>(columnName: 'name', value: newCellValue);
      employees[dataRowIndex].name = newCellValue.toString();
    } else if (column.columnName == 'designation') {
      dataGridRows[dataRowIndex].getCells()[rowColumnIndex.columnIndex] =
          DataGridCell<String>(columnName: 'designation', value: newCellValue);
      employees[dataRowIndex].designation = newCellValue.toString();
    } else {
      dataGridRows[dataRowIndex].getCells()[rowColumnIndex.columnIndex] =
          DataGridCell<int>(columnName: 'salary', value: newCellValue);
      employees[dataRowIndex].salary = newCellValue as int;

      if (dataRowIndex == dataGridRows.length - 1) {
        DataGridRow dataGridRow = DataGridRow(cells: <DataGridCell>[
          DataGridCell<int>(columnName: 'id', value: 1),
          DataGridCell<String>(columnName: 'name', value: 'Perry'),
          DataGridCell<String>(columnName: 'designation', value: 'Developer'),
          DataGridCell<int>(columnName: 'salary', value: 45000),
        ]);
        dataGridRows.add(dataGridRow);

        Employee employee = Employee(1, 'Perry', 'Developer', 45000);
        employees.add(employee);
        notifyListeners();

        WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
          dataGridController.beginEdit(dataGridController.currentCell);
        });
      }
    }
  }

  // Navigate Function to get value from next page
  void navigateToSearch (){
    isSearching = true;
    Navigator.of(context).push(MaterialPageRoute(builder: (context) => SearchPage())).then((value)
    {
      isSearching = false;
      valueChange = false;
    }
    );

  }

  @override
  bool canSubmitCell(DataGridRow dataGridRow, RowColumnIndex rowColumnIndex, GridColumn column) {

    if (valueChange == true)
    {
      if(isSearching == false)
      {
        navigateToSearch();
      }
      return true;
    }
    else
    {
     return true;
    }

    // TODO: implement canSubmitCell
    return super.canSubmitCell(dataGridRow, rowColumnIndex, column);
  }

  @override
  Widget? buildEditWidget(DataGridRow dataGridRow,
      RowColumnIndex rowColumnIndex, GridColumn column, CellSubmit submitCell) {
    // Text going to display on editable widget
    final String displayText = dataGridRow
        .getCells()
        .firstWhereOrNull((DataGridCell dataGridCell) =>
    dataGridCell.columnName == column.columnName)
        ?.value
        ?.toString() ??
        '';

    // The new cell value must be reset.
    // To avoid committing the [DataGridCell] value that was previously edited
    // into the current non-modified [DataGridCell].
    newCellValue = null;

    final bool isNumericType =
        column.columnName == 'id' || column.columnName == 'salary';

    // Holds regular expression pattern based on the column type.
    final RegExp regExp = _getRegExp(isNumericType, column.columnName);

    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      alignment: isNumericType ? Alignment.centerRight : Alignment.centerLeft,
      child: TextField(
        style: const TextStyle(fontSize: 14),
        controller: editingController..text = displayText,
        textAlign: isNumericType ? TextAlign.right : TextAlign.left,
        autofocus: true,
        decoration: const InputDecoration(
          border: InputBorder.none,
        ),
        inputFormatters: <TextInputFormatter>[
          FilteringTextInputFormatter.allow(regExp)
        ],
        keyboardType: isNumericType ? TextInputType.number : TextInputType.text,
        onChanged: (String value) {
          valueChange  = true;
          if (value.isNotEmpty) {
            if (isNumericType) {
              newCellValue = int.parse(value);
            } else {
              newCellValue = value;
            }
          } else {
            newCellValue = null;
          }
        },
        onSubmitted: (String value) {
          /// Call [CellSubmit] callback to fire the canSubmitCell and
          /// onCellSubmit to commit the new value in single place.
          submitCell();
        },
      ),
    );
  }

  RegExp _getRegExp(bool isNumericKeyBoard, String columnName) {
    return isNumericKeyBoard ? RegExp('[0-9]') : RegExp('[a-zA-Z ]');
  }
}

class Employee {
  Employee(this.id, this.name, this.designation, this.salary);

  int? id;
  String? name;
  String? designation;
  int? salary;

  DataGridRow getDataGridRow() {
    return DataGridRow(cells: <DataGridCell>[
      DataGridCell<int>(columnName: 'id', value: id),
      DataGridCell<String>(columnName: 'name', value: name),
      DataGridCell<String>(columnName: 'designation', value: designation),
      DataGridCell<int>(columnName: 'salary', value: salary),
    ]);
  }
}

///secound page for searching data
class SearchPage extends StatefulWidget {
  /// Creates the Search page.
  SearchPage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<SearchPage> {
  List<Employee> employees = <Employee>[];
  late SearchDataSource searchDataSource;

  @override
  void initState() {
    super.initState();
    employees = getEmployeeData();
    searchDataSource = SearchDataSource(employeeData: employees);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Syncfusion Flutter DataGrid'),
      ),
      body: SfDataGrid(
        source: searchDataSource,
        columnWidthMode: ColumnWidthMode.fill,
        columns: <GridColumn>[
          GridColumn(
              columnName: 'id',
              label: Container(
                  padding: EdgeInsets.all(16.0),
                  alignment: Alignment.center,
                  child: Text(
                    'ID',
                  ))),
          GridColumn(
              columnName: 'name',
              label: Container(
                  padding: EdgeInsets.all(8.0),
                  alignment: Alignment.center,
                  child: Text('Name'))),
          GridColumn(
              columnName: 'designation',
              label: Container(
                  padding: EdgeInsets.all(8.0),
                  alignment: Alignment.center,
                  child: Text(
                    'Designation',
                    overflow: TextOverflow.ellipsis,
                  ))),
          GridColumn(
              columnName: 'salary',
              label: Container(
                  padding: EdgeInsets.all(8.0),
                  alignment: Alignment.center,
                  child: Text('Salary'))),
        ],
      ),
    );
  }

  List<Employee> getEmployeeData() {
    return [
      Employee(10001, 'James', 'Project Lead', 20000),
      Employee(10002, 'Kathryn', 'Manager', 30000),
      Employee(10003, 'Lara', 'Developer', 15000),
      Employee(10004, 'Michael', 'Designer', 15000),
      Employee(10005, 'Martin', 'Developer', 15000),
      Employee(10006, 'Newberry', 'Developer', 15000),
      Employee(10007, 'Balnc', 'Developer', 15000),
      Employee(10008, 'Perry', 'Developer', 15000),
      Employee(10009, 'Gable', 'Developer', 15000),
      Employee(10010, 'Grimes', 'Developer', 15000)
    ];
  }
}

/// An object to set the employee collection data source to the datagrid. This
/// is used to map the employee data to the datagrid widget.
class SearchDataSource extends DataGridSource {
  /// Creates the employee data source class with required details.
  SearchDataSource({required List<Employee> employeeData}) {
    _employeeData = employeeData
        .map<DataGridRow>((e) => DataGridRow(cells: [
      DataGridCell<int>(columnName: 'id', value: e.id),
      DataGridCell<String>(columnName: 'name', value: e.name),
      DataGridCell<String>(
          columnName: 'designation', value: e.designation),
      DataGridCell<int>(columnName: 'salary', value: e.salary),
    ]))
        .toList();
  }

  List<DataGridRow> _employeeData = [];

  @override
  List<DataGridRow> get rows => _employeeData;

  @override
  DataGridRowAdapter buildRow(DataGridRow row) {
    return DataGridRowAdapter(
        cells: row.getCells().map<Widget>((e) {
          return Container(
            alignment: Alignment.center,
            padding: EdgeInsets.all(8.0),
            child: Text(e.value.toString()),
          );
        }).toList());
  }
}

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @abdulrehmananwar,

We are not clear on what you are trying to achieve. Based on the sample you have provided, there is another DataGrid present on the SearchPage. **Are you trying to navigate to the cells within the rows of the DataGrid on the SearchPage using the arrow keys (up/down) and the tab key on your keyboard? Could you please provide an additional details and clarification on your requirements for a better understanding?**

Regards,
Tamilarasan

**Comment by abdulrehmananwar** [comment]
basically when user input some value this will open another gridview from where he will choose value as per input cell. and on selection secound gridview will be close and selected value will be punch to parent gridview.

**Comment by abdulrehmananwar** [comment]
please update on it

**Comment by Tamilarasan-Paranthaman** [comment]
@abdulrehmananwar,

Based on the information and sample provided, it seems that if you navigate to a new page in the canSubmitCell method, the method will not wait until a value is selected from the new page. As a result, the cell value will be committed, triggering the respective tab or up/down key navigation from the first DataGrid. Additionally, the first DataGrid was not disposed when navigating to the new page, which is why the focus was received on the second page. To execute the remaining code in the method after returning from the new page, you can use the await **Navigator.push** method to navigate to the new page and **await** the result. We have already considered the feature of making the **canSubmitCell** method as async. We suspect that this feature can satisfy your requirements. We will implement this feature in our 2023 Volume 1 release, which is expected to be rolled out on end of March 2023. You can follow up with the below feedback for further details,

Feedback Link: https://www.syncfusion.com/feedback/33165/make-oncellsubmit-method-as-async-to-process-the-asynchronous-operation-on-editing 

As of now, you can achieve your requirement in TextField.onChanged callback in the buildEditWidget method. Please check the following code snippet. We also want to note that in your provided sample, the selection mode and navigation mode were not set as cell properties in the SearchPage(second DataGrid). Make sure to set these properties to enable navigation.

```dart
@override
  Widget? buildEditWidget(DataGridRow dataGridRow,
      RowColumnIndex rowColumnIndex, GridColumn column, CellSubmit submitCell) {

…

    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      alignment: isNumericType ? Alignment.centerRight : Alignment.centerLeft,
      child: TextField(
        style: const TextStyle(fontSize: 14),
        controller: editingController..text = displayText,
        textAlign: isNumericType ? TextAlign.right : TextAlign.left,
        autofocus: true,
        decoration: const InputDecoration(
          border: InputBorder.none,
        ),
        inputFormatters: <TextInputFormatter>[
          FilteringTextInputFormatter.allow(regExp)
        ],
        keyboardType: isNumericType ? TextInputType.number : TextInputType.text,
        onChanged: (String value) async {
          valueChange = true;

          if (value.isNotEmpty) {
            if (value == 'emp') {
              if (valueChange == true) {
                if (isSearching == false) {
                  navigateToSearch();
                }
              }
            } else {
              if (isNumericType) {
                newCellValue = int.parse(value);
              } else {
                newCellValue = value;
              }
            }
          } else {
            newCellValue = null;
          }
        },
        onSubmitted: (String value) {
          /// Call [CellSubmit] callback to fire the canSubmitCell and
          /// onCellSubmit to commit the new value in single place.
          submitCell();
        },
      ),
    );
  }
  }

```

Regards,
Tamilarasan

**Comment by abdulrehmananwar** [comment]
in this way on each word typing this will navigateToSearch. butt i want to navigateToSearch after completion on sentence.  currenty i am working on cell submit function butt there i cant check cansubmit. can if user input wrong sentence i want to force it on current cell.

**Comment by Tamilarasan-Paranthaman** [comment]
@abdulrehmananwar ,

We are glad to announce that our Essential Studio 2023 Volume 1 Main Release V21.1.0.35 is rolled out. The support for making the onCellSubmit method async to process the asynchronous operation on editing completion has been included in version **21.1.35**.

Pub Link: https://pub.dev/packages/syncfusion_flutter_datagrid/changelog

UG Documentation: https://help.syncfusion.com/flutter/datagrid/editing#perform-editing-asynchronously

We thank you for your support and appreciate your patience in waiting for this release. Please get in touch with us if you would require any further assistance.
